feat: Vertex AI Search Tuning sample notebook

JincheolKim opened the pull request:
Contributing an additional notebook for the "Vertex AI Search Tuning" feature to help users try it with a sample notebook.

---

- [V] Follow the [`CONTRIBUTING` Guide](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/CONTRIBUTING.md).
- [V] You are listed as the author in your notebook or README file.
  - [ ] Your account is listed in [`CODEOWNERS`](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/.github/CODEOWNERS) for the file(s).
- [V] Make your Pull Request title in the <https://www.conventionalcommits.org/> specification.
- [V] Ensure the tests and linter pass (Run `nox -s format` from the repository root to format).
- [V] Appropriate docs were updated (if necessary)

No issue fix for this PR.

---

JincheolKim commented:
Would you give me a guidance to resolve the failures from

Check spelling
Lint Code Base

checks during the PR process? Thank you very much.

holtskinner reviewed (changes requested):
Remove all of the duplicate files.

Move the Notebook to a folder called `search/tuning`.

Title it `vertex-ai-search-tuning.ipynb`

Change the Image link to be `https://storage.googleapis.com/github-repo/generative-ai/search/tuning/images/key_user_journey_search_tuning.png`

Remove all of the tuning data from this PR, and use the data from this public GCS Bucket `gs://github-repo/generative-ai/search/tuning/tuning_data/` Either have the users pull directly from this Bucket for the tuning job, or have them create the bucket and do `gsutil cp` to copy the data into their own bucket from this one.

Also, be sure to fix the spelling and lint errors

JincheolKim commented:
Please create a subfolder named "awesome-rlhf" under the "gs://github-repo/generative-ai/search/tuning" subfolder. I will move all of the data under the public bucket. Thank you.

JincheolKim commented:
> gs://github-repo/generative-ai/search/tuning

Please create a subfolder named "awesome_rlhf" under the "gs://github-repo/generative-ai/search/tuning" subfolder or would you be able to give me the ""Create/Delete/Read/Write" right to the "gs://github-repo/generative-ai/search/tuning" folder to create the subfolder? I need to move all of the test data under the subfolder to make the asset complied with your comments.

JincheolKim commented:
All of the requested corrections and actions are completed and committed. Please review them again. Thank you.

holtskinner commented on .github/actions/spelling/excludes.txt:
```diff
@@ -114,4 +114,5 @@ ignore$
 ^\Qowlbot.py\E$
 ^\Q.github/workflows/issue_assigner/assign_issue.py\E$
 ^\Qnoxfile.py\E$
+^\Qsearch/tuning/vertexai-search-tuning.ipynb\E$
```
```suggestion
```

holtskinner commented on search/tuning/vertexai-search-tuning.ipynb:
```diff
@@ -0,0 +1,1637 @@
+        ") -> str:\n",
+        "    #  For more information, refer to:\n",
+        "    # https://cloud.google.com/generative-ai-app-builder/docs/locations#specify_a_multi-region_for_your_data_store\n",
+        "    client_options = (\n",
+        "        ClientOptions(api_endpoint=f\"{location}-discoveryengine.googleapis.com\")\n",
+        "        if location != \"global\"\n",
+        "        else None\n",
+        "    )\n",
+        "\n",
```
I'd recommend making a single `ClientOptions` object that all of these methods can use.

Also, move helper functions to a dedicated section at the top so that the rest of the notebook can just focus on the key steps, not the full code needed.

JincheolKim replied:
`client_option` is defined once at the "Helper functions to facilitate the following steps" section in the earlier part of the notebook.

holtskinner commented on search/tuning/vertexai-search-tuning.ipynb:
```diff
@@ -0,0 +1,1637 @@
+      "outputs": [],
+      "source": [
+        "! gcloud storage buckets create --location={STORAGE_LOCATION} --project={PROJECT_ID} --enable-hierarchical-namespace --uniform-bucket-level-access -b {BUCKET_URI}\n",
+        "! mkdir $TUNING_DATA_PATH_LOCAL\n",
+        "! gcloud storage cp $TUNING_DATA_PATH_SOURCE/* $TUNING_DATA_PATH_LOCAL"
```
Why is this data being copied locally? Is that absolutely needed? Could the GCS path be given directly to VAIS?

JincheolKim replied:
These data are used to extract the corpus and the query JSONL data, the training and the test TSV data locally. Given that, those files are downloaded to a local folder, preprocessed, and then uploaded to the data store again. This logic is a correct logic.

holtskinner commented on search/tuning/vertexai-search-tuning.ipynb:
```diff
@@ -0,0 +1,1637 @@
+      },
+      "outputs": [],
+      "source": [
+        "!gcloud storage cp \"{TUNING_DATA_PATH_LOCAL}/*.pdf\" \"{SEARCH_DATASTORE_PATH_REMOTE}\"\n",
+        "!gcloud storage ls \"{SEARCH_DATASTORE_PATH_REMOTE}\""
```
Similar to above, do these files need to be copied locally?

JincheolKim replied:
Same as in the above. This part is to upload the PDF files corresponding to the JSONL and the TSV format files to generate the index in addition to the base dataset.

holtskinner commented on search/tuning/vertexai-search-tuning.ipynb:
```diff
@@ -0,0 +1,1637 @@
+        "from google.cloud import discoveryengine_v1 as discoveryengine\n",
+        "\n",
+        "\n",
+        "def delete_engine(\n",
+        "    project_id: str,\n",
+        "    location: str,\n",
+        "    engine_id: str,\n",
+        ") -> str:\n",
```
Same as above, move these helper methods to a dedicated section at the top

JincheolKim replied:
This "delete_engine" helper function was also moved to the section "Helper Functions".

JincheolKim commented on search/tuning/vertexai-search-tuning.ipynb:
```diff
@@ -0,0 +1,1637 @@
+        "import re\n",
+        "\n",
+        "\n",
+        "def generate_source_dataset(\n",
+        "    source_file, corpus_filepath, query_filepath, cleanup_at_start=True\n",
+        "):\n",
```
Added descriptions on the helper function arguments, return values, and their functions.

JincheolKim commented on search/tuning/vertexai-search-tuning.ipynb:
```diff
@@ -0,0 +1,1637 @@
+        "def generate_source_dataset(\n",
+        "    source_file, corpus_filepath, query_filepath, cleanup_at_start=True\n",
+        "):\n",
+        "    import logging\n",
+        "    import os\n",
```
All of the imports were aggregated to a cell in the section "Set Google Cloud project information and initialize Vertex AI SDK for Python".

JincheolKim commented on search/tuning/vertexai-search-tuning.ipynb:
```diff
@@ -0,0 +1,1637 @@
+      },
+      "outputs": [],
+      "source": [
+        "# For Linux\n",
+        "!pandoc --pdf-engine=pdflatex FAQ-Kubernetes-Client.md -o FAQ-Kubernetes-Client.pdf\n",
+        "!pandoc --pdf-engine=pdflatex FAQ.md -o FAQ.pdf\n",
+        "!pandoc --pdf-engine=pdflatex README.md -o README.pdf\n",
+        "## For macOS\n",
+        "#!pandoc --pdf-engine=xelatex FAQ-Kubernetes-Client.md -o FAQ-Kubernetes-Client.pdf\n",
+        "#!pandoc --pdf-engine=xelatex FAQ-Kubernetes-Client.md -o FAQ-Kubernetes-Client.pdf\n",
+        "#!pandoc --pdf-engine=xelatex FAQ-Kubernetes-Client.md -o FAQ-Kubernetes-Client.pdf"
```
Added a conditional branch statements with the value of the platform.system() function.

JincheolKim commented on search/tuning/vertexai-search-tuning.ipynb:
```diff
@@ -0,0 +1,1637 @@
+      },
+      "outputs": [],
+      "source": [
+        "!git clone https://gitlab.com/jincheolkim/awesome-rlhf.git\n",
```
Corrected as you suggested. All of the data will be downloaded from the GCS bucket.

JincheolKim commented on search/tuning/vertexai-search-tuning.ipynb:
```diff
@@ -0,0 +1,1637 @@
+      },
+      "outputs": [],
+      "source": [
+        "from google.api_core.client_options import ClientOptions\n",
+        "from google.cloud import discoveryengine\n",
```
All of the imports were aggregated to a cell in the section "Set Google Cloud project information and initialize Vertex AI SDK for Python".

JincheolKim commented on search/tuning/vertexai-search-tuning.ipynb:
```diff
@@ -0,0 +1,1637 @@
+      },
+      "outputs": [],
+      "source": [
+        "# Create a client\n",
+        "client = discoveryengine.DocumentServiceClient(client_options=client_options)"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "33e54f26212a"
+      },
+      "outputs": [],
+      "source": [
+        "# The full resource name of the search engine branch.\n",
+        "# e.g. projects/{project}/locations/{location}/dataStores/{data_store_id}/branches/{branch}\n",
+        "parent = client.branch_path(\n",
+        "    project=PROJECT_ID,\n",
+        "    location=LOCATION,\n",
+        "    data_store=SEARCH_DATASTORE_ID,\n",
```
The logic was implemented as the function "import_documents" as other helper functions in the "Helper Functions" section.

JincheolKim commented:
All of the requested fix are corrected and revised. Your review and merge would be greatly appreciated. 

---
1. Added descriptions on the helper function arguments, return values, and their functions. (generate_source_dataset, generate_training_test_dataset)
2. All of the imports were aggregated to a cell in the section "Set Google Cloud project information and initialize Vertex AI SDK for Python".
3. In the cell executing pandoc to transform .md text files to PDF files, a conditional branch statement with the value of the platform.system() function was added.
4. Parts which are downloading from a personal github repository were all removed and corrected to be downloaded from the GCS bucket, gs://github-repo/generative-ai/search/tuning.
5. All of the imports were aggregated to a cell in the section "Set Google Cloud project information and initialize Vertex AI SDK for Python".
6. client_option is defined once at the "Helper functions to facilitate the following steps" section in the earlier part of the notebook.
7. The logic in the section “Creating a data store for a search app with the cloud storage bucket with PDF documents” was implemented as the function "import_documents" as other helper functions in the "Helper Functions" section.
8. This "delete_engine" helper function was also moved to the section "Helper Functions".
